services/sso/src/main/java/software.amazon.awssdk.services.sso/auth/SsoCredentialsProvider.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.sso.auth;
+
+import static software.amazon.awssdk.utils.Validate.notNull;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.auth.credentials.SessionCredentialsHolder;
+import software.amazon.awssdk.profiles.ProfileFile;
+import software.amazon.awssdk.services.sso.SsoClient;
+import software.amazon.awssdk.services.sso.internal.ExpiredTokenException;
+import software.amazon.awssdk.services.sso.model.GetRoleCredentialsRequest;
+import software.amazon.awssdk.services.sso.model.RoleCredentials;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+import software.amazon.awssdk.utils.cache.CachedSupplier;
+import software.amazon.awssdk.utils.cache.NonBlocking;
+import software.amazon.awssdk.utils.cache.RefreshResult;
+
+/**
+ * An implementation of {@link AwsCredentialsProvider} that is extended within this package to provide support for periodically
+ * updating session credentials. The credential provider loads configuration from a {@link ProfileFile}, and use this information
+ * to generate a cached token directory. With this directory the information a {@link SsoClient} can do a
+ * {@link SsoClient#getRoleCredentials(Consumer)} call to retrieve the credentials needed.
+ *
+ * Currently, the cached token is assumed unexpired, and if it's expired then an {@link ExpiredTokenException} will be thrown. If
+ * the users want to change the behavior of this, please implement your own cached token resolving logic and override the
+ * {@link Builder#refreshRequest).
+ *
+ * When credentials get close to expiration, this class will attempt to update them asynchronously. If the credentials
+ * end up expiring, this class will block all calls to {@link #resolveCredentials()} until the credentials can be updated.
```

Comment:
Is this still accurate? E.g. with regard loading from a profile file.

Reply:
Generally, we still need to load the `accountId` and `roleName` from the profile file, but that's for generating the request, not the cached token directory. I can update that.